security/pkg/nodeagent/caclient/providers/citadel/client.go

```diff
@@ -162,3 +162,39 @@ func (c *citadelClient) getTLSDialOption() (grpc.DialOption, error) {
 	transportCreds := credentials.NewTLS(&config)
 	return grpc.WithTransportCredentials(transportCreds), nil
 }
+
+func (c *citadelClient) buildConnection() (*grpc.ClientConn, error) {
+	var opts grpc.DialOption
+	var err error
+	if c.enableTLS {
+		opts, err = c.getTLSDialOption()
+		if err != nil {
+			return nil, err
+		}
+	} else {
+		opts = grpc.WithInsecure()
+	}
+
+	conn, err := grpc.Dial(c.caEndpoint, opts)
+	if err != nil {
+		citadelClientLog.Errorf("Failed to connect to endpoint %s: %v", c.caEndpoint, err)
+		return nil, fmt.Errorf("failed to connect to endpoint %s", c.caEndpoint)
+	}
+
+	return conn, nil
+}
+
+func (c *citadelClient) reconnect() error {
+	err := c.conn.Close()
+	if err != nil {
+		return fmt.Errorf("failed to close connection")
```

Comment:
By looking at Close(), it only returns err when the inner conns is nil. Maybe when Close() returns err, we can still proceed to create a new conn, with low risk of any resource leak.

Reply:
inner conns should not be nil. Because every time we reconnect and the first time we create the client we will set the c.conn to that connection that citadel client are using it.